### PR TITLE
Fixes NER to correctly expand/shrink the labels

### PR DESCRIPTION
**Pull request by michaelgsharp**
NER now will correctly expand/shrink the labels when a word is split into multiple tokens.

### CODE REVIEW

File: test/Microsoft.ML.Tests/NerTests.cs [hunk 1]
```diff
@@ -36,25 +38,115 @@ public void TestSimpleNer()
...
                         Label = new string[]{"PERSON", "0", "PERSON", "0", "0", "0", "COUNTRY"}
+                     },
+                     new()
+                     {
+                         Sentence = "WW II Landmarks on the Great Earth of China : Eternal Memories of Taihang Mountain",
+                         Label = new string[]{"B_WORK_OF_ART", "WORK_OF_ART", "WORK_OF_ART", "WORK_OF_ART", "WORK_OF_ART", "WORK_OF_ART", "WORK_OF_ART", "WORK_OF_ART", "WORK_OF_ART", "WORK_OF_ART", "WORK_OF_ART", "WORK_OF_ART", "WORK_OF_ART", "WORK_OF_ART", "WORK_OF_ART" }
+                     },
+                     new()
+                     {
+                         Sentence = "This campaign broke through the Japanese army 's blockade to reach base areas behind enemy lines , stirring up anti-Japanese spirit throughout the nation and influencing the situation of the anti-fascist war of the people worldwide .",
```
Comment (jeffhandley) [inline test/Microsoft.ML.Tests/NerTests.cs:66]: Should the space in `army 's` be removed?

File: src/Microsoft.ML.Tokenizers/Model/EnglishRoberta.cs [hunk 2]
```diff
@@ -27,6 +27,7 @@ public sealed class EnglishRoberta : Model
         private readonly IReadOnlyDictionary<char, char> _unicodeToByte;
         private readonly string[] _charToString;
         private readonly Cache<string, IReadOnlyList<Token>> _cache;
+        private readonly char _startChar = '\u0120';
```
Comment (ericstj) [inline src/Microsoft.ML.Tokenizers/Model/EnglishRoberta.cs:30]:
```suggestion
        private const char StartChar = (char)(' ' + 256);
```
I found this a helpful explanation of why this is so: https://github.com/openai/gpt-2/issues/80#issuecomment-487202159

Seems like we have a similar function here which does the 256 shift: https://github.com/dotnet/machinelearning/blob/51d8dc47480743b8c30b8622e3c3f094d76c44ef/src/Microsoft.ML.Tokenizers/Model/EnglishRoberta.cs#L458-L482

Might be good to explain how these are connected.

File: src/Microsoft.ML.TorchSharp/NasBert/NerTrainer.cs [hunk 3]
```diff
@@ -334,6 +370,41 @@ private protected override Delegate CreateGetter(DataViewRow input, int iinfo, T
 
             }
 
+            private void CondenseOutput(ref VBuffer<UInt32> dst, string sentence, Tokenizer tokenizer, TensorCacher outputCacher)
+            {
+                var pre = tokenizer.PreTokenizer.PreTokenize(sentence);
+                TokenizerResult encoding = tokenizer.Encode(sentence);
```
Comment (ericstj) [inline src/Microsoft.ML.TorchSharp/NasBert/NerTrainer.cs:376]: Does this tokenize the same sentence that was previously tokenized as input to the model?

File: test/Microsoft.ML.Tests/NerTests.cs (hunk 1 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.Tests/NerTests.cs:66]: Nope. That is correct for how the input labels are put. Trying to test edge cases with stuff like that.

File: src/Microsoft.ML.Tokenizers/Model/EnglishRoberta.cs (hunk 2 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Tokenizers/Model/EnglishRoberta.cs:30]: We do have a function that does a 256 shift, but this is specifically shifting the "start" token so we can use it for comparisons and stuff. I'll add a note about it though.

File: src/Microsoft.ML.TorchSharp/NasBert/NerTrainer.cs (hunk 3 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.TorchSharp/NasBert/NerTrainer.cs:376]: It does. I didn't change it to use what was already encoded cause of what I would have to change (base classes and other classes that use the base class) and that this isn't time consuming at all compared to the actual processing of the input. I am happy to make the changes though.

File: src/Microsoft.ML.Tokenizers/Model/Model.cs [hunk 4]
```diff
@@ -67,6 +67,12 @@ public abstract class Model
         /// <returns></returns>
         public abstract bool IsValidChar(char ch);
 
+        /// <summary>
+        /// Returns if the first character of the token is part of the actual word or not.
+        /// </summary>
+        /// <param name="token"></param>
+        /// <returns></returns>
+        public abstract bool IsFirstTokenInWord(string token);
```
Comment (tarekgh) [inline src/Microsoft.ML.Tokenizers/Model/Model.cs:75]: >IsFirstTokenInWord [](http://example.com/codeflow?start=29&length=18)

Do we know if this is going to be useful for all tokenizers models? I am asking to know if we move this to the NasBert which know it is using EnglishRoberta tokenizer. I think we need to add to the Model only the functionality we believe will benefit most tokenizers.

File: src/Microsoft.ML.Tokenizers/Model/EnglishRoberta.cs [hunk 5]
```diff
@@ -592,6 +593,13 @@ public override bool IsValidChar(char ch)
         {
             return _byteToUnicode.ContainsKey(ch);
         }
+
+        public override bool IsFirstTokenInWord(string token)
```
Comment (tarekgh) [inline src/Microsoft.ML.Tokenizers/Model/EnglishRoberta.cs:597]: >public override bool IsFirstTokenInWord(string token) [](http://example.com/codeflow?start=8&length=53)

nit:

public override bool IsFirstTokenInWord(string token) => token is null || (token.Length != 0 && token[0] == _startChar)

File: src/Microsoft.ML.TorchSharp/NasBert/NasBertTrainer.cs [hunk 6]
```diff
@@ -583,13 +584,13 @@ private IList<int> PrepInputTokens(ref ReadOnlyMemory<char> sentence1, ref ReadO
                 getSentence1(ref sentence1);
                 if (getSentence2 == default)
                 {
-                    return new[] { 0 /* InitToken */ }.Concat(tokenizer.EncodeToConverted(sentence1.ToString())).ToList();
+                    return InitTokenArray.Concat(tokenizer.EncodeToConverted(sentence1.ToString())).ToList();
```
Comment (tarekgh) [inline src/Microsoft.ML.TorchSharp/NasBert/NasBertTrainer.cs:587]: >return InitTokenArray.Concat(tokenizer.EncodeToConverted(sentence1.ToString())).ToList(); [](http://example.com/codeflow?start=20&length=89)

nit:

If you want to save one more allocation here, you may do something like:

```C#
List<int> newList = new List<int>(tokenizer.EncodeToConverted(sentence1.ToString()));
newList.Insert(0, 0);
return newList;
```

You don't have to do it if it is not worth it.

File: src/Microsoft.ML.TorchSharp/NasBert/NerTrainer.cs [hunk 7]
```diff
@@ -155,6 +157,40 @@ private protected override torch.Tensor CreateTargetsTensor(ref List<TargetType>
...
+                var encoding = Tokenizer.Encode(sentence);
+
+                if (target.Length != encoding.Tokens.Count)
+                {
+                    var targetIndex = 0;
+                    var targetEditor = VBufferEditor.Create(ref target, encoding.Tokens.Count);
+                    var newValues = targetEditor.Values;
+                    for (var i = 0; i < encoding.Tokens.Count; i++)
+                    {
+                        if (Tokenizer.Model.IsFirstTokenInWord(encoding.Tokens[i]))
```
Comment (tarekgh) [inline src/Microsoft.ML.TorchSharp/NasBert/NerTrainer.cs:175]: >IsFirstTokenInWord [](http://example.com/codeflow?start=44&length=18)

I think the name of this method is not accurate because you can get a work with starting token not having `Ġ`. 
here is example I found:
```
text = '(Besides that there should be more restaurants like it around the city).'
Token (str): ['<s>', 'Ġ(', 'Besides', 'Ġthat', 'Ġthere', 'Ġshould', 'Ġbe', 'Ġmore', 'Ġrestaurants', 'Ġlike', 'Ġit', 'Ġaround', 'Ġthe', 'Ġcity', ')', 'Ġ.', '</s>']
```

I would instead call this method as TokenStartsWithSpace.

Comment (michaelgsharp) [reply src/Microsoft.ML.TorchSharp/NasBert/NerTrainer.cs:175]: Good point. Thats how it was called in the MSR code, but I think your name makes more sense here. I'll get it updated.

Comment (ericstj) [reply src/Microsoft.ML.TorchSharp/NasBert/NerTrainer.cs:175]: From that perspective - would @tarekgh's example break our scenario?  What if a named entity is proceeded with a `(` - how does the model identify that?

File: src/Microsoft.ML.TorchSharp/NasBert/NerTrainer.cs (hunk 3 above)
Comment (ericstj) [reply src/Microsoft.ML.TorchSharp/NasBert/NerTrainer.cs:376]: How does the implementation we ported this from handle the same scenario?  Does it recalculate the tokenization multiple times or does it somehow flow it through?
We could always keep this the way it is, and come back to fix it if the perf matters, though I was wondering if we somehow flowed more information about the tokenization through the data we might be able to avoid having to expose new API on the tokenizer.

File: src/Microsoft.ML.Tokenizers/Model/Model.cs (hunk 4 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Tokenizers/Model/Model.cs:75]: We could just move it to NasBert I think. Let me do that.

File: src/Microsoft.ML.TorchSharp/NasBert/NasBertTrainer.cs (hunk 6 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.TorchSharp/NasBert/NasBertTrainer.cs:587]: Changed it to that.

File: src/Microsoft.ML.TorchSharp/NasBert/NasBertTrainer.cs [hunk 8]
```diff
@@ -583,13 +584,15 @@ private IList<int> PrepInputTokens(ref ReadOnlyMemory<char> sentence1, ref ReadO
                 getSentence1(ref sentence1);
                 if (getSentence2 == default)
                 {
-                    return new[] { 0 /* InitToken */ }.Concat(tokenizer.EncodeToConverted(sentence1.ToString())).ToList();
+                    List<int> newList = new List<int>(tokenizer.EncodeToConverted(sentence1.ToString()));
+                    newList.Insert(0, 0);
```
Comment (tarekgh) [inline src/Microsoft.ML.TorchSharp/NasBert/NasBertTrainer.cs:588]: >newList.Insert(0, 0); [](http://example.com/codeflow?start=20&length=21)

worth a comment telling insert InitToken.

**Review by tarekgh — APPROVED** [review]

File: src/Microsoft.ML.TorchSharp/NasBert/NasBertTrainer.cs (hunk 8 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.TorchSharp/NasBert/NasBertTrainer.cs:589]: Just added in a comment for it.
